tests/e2e/tests/pilot/performance/serviceentry_test.go

```diff
@@ -0,0 +1,328 @@
+package performance
+
+import (
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"math/rand"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/types"
+	"github.com/google/uuid"
+	"github.com/onsi/gomega"
+
+	mcp "istio.io/api/mcp/v1alpha1"
+	networking "istio.io/api/networking/v1alpha3"
+	mixerEnv "istio.io/istio/mixer/test/client/env"
+	"istio.io/istio/pilot/pkg/bootstrap"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/adsc"
+	"istio.io/istio/pkg/mcp/snapshot"
+	mcptest "istio.io/istio/pkg/mcp/testing"
+	"istio.io/istio/tests/util"
+)
+
+var (
+	serviceEntryCount = flag.Int("serviceEntryCount", 1000, "Number of Service Entries to register")
+	terminateAfter    = flag.Int("terminateAfter", 60, "Number of seconds to run the test")
+	envoyCount        = flag.Int("adsc", 1, "Number of concurrent envoys to connect to pilot")
+)
+
+const (
+	pilotDebugPort = 5555
+	pilotGrpcPort  = 15010
+	serviceDomain  = "example.com"
+)
+
+func TestServiceEntry(t *testing.T) {
+	t.Logf("running performance test with the following parameters { Number of ServiceEntry: %d, Number of Envoys: %d, Total test time: %ds }",
+		*serviceEntryCount,
+		*envoyCount,
+		*terminateAfter)
+
+	g := gomega.NewGomegaWithT(t)
+
+	t.Log("building & starting mock mcp server...")
+	mcpServer, err := runMcpServer()
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+	defer mcpServer.Close()
+
+	go runSnapshot(mcpServer)
+
+	debugAddr := fmt.Sprintf("127.0.0.1:%d", pilotDebugPort)
+	pilotAddr := fmt.Sprintf("127.0.0.1:%d", pilotGrpcPort)
+
+	tearDown := initLocalPilotTestEnv(t, mcpServer.Port, pilotAddr, debugAddr)
+	defer tearDown()
+
+	t.Log("run sidecar envoy(s)...")
+	adscs := adsConnectAndWait(*envoyCount, pilotAddr)
+	for _, adsc := range adscs {
+		defer adsc.Close()
+	}
+
+	t.Log("check for service entries in pilot's debug endpoint...")
+	g.Eventually(func() (int, error) {
+		return registeredServiceEntries(fmt.Sprintf("http://127.0.0.1:%d/debug/configz", pilotDebugPort))
+	}, "30s", "1s").Should(gomega.Equal(*serviceEntryCount))
+
+	t.Log("check for registered endpoints in envoys's debug endpoint...")
+	g.Eventually(func() error {
+		for _, adsc := range adscs {
+			var data map[string]interface{}
+			err = json.Unmarshal([]byte(adsc.EndpointsJSON()), &data)
+			if err != nil {
+				return err
+			}
+			if len(data) != *serviceEntryCount {
+				return errors.New("endpoint not registered")
+			}
+		}
+		return nil
+	}, "30s", "1s").ShouldNot(gomega.HaveOccurred())
+
+	terminate()
+}
+
+func runSnapshot(mcpServer *mcptest.Server) {
+	configInterval := time.NewTicker(time.Second * 3)
+	v := 0
+	b := snapshot.NewInMemoryBuilder()
+	for {
```

Comment:
How does this loop exit?

Reply:
Initially it had an exit strategy but it seems to have gotten removed during refactor and never got added back in, it's fixed now.